src/tests/playwright/pages/util/recorder/recorder-main.js

```diff
@@ -0,0 +1,50 @@
+/**
+ * @fileoverview AudioWorklet Recorder Node to record raw PCM data from
+ * Web Audio Graph.
+ */
+
+/**
+ * Creates an AudioWorklet recorder node to record input for a specified
+ * length of time (seconds). Passes audio through to output.
+ * @param {AudioContext} context - The AudioContext to create the recorder node.
+ * @param {number} recordLength - The length in seconds to record for.
+ * @return {Object} An object containing:
+ *   - {AudioWorkletNode} recorder: The recorder AudioWorkletNode.
+ *   - {Promise<AudioBuffer>} bufferPromise: A promise that resolves to an
+ *   AudioBuffer once filled.
+ */
+export const record = async (context, recordLength) => {
+  console.assert(context instanceof AudioContext);
+  console.assert(typeof recordLength === 'number' && recordLength > 0);
+
+  const maxSamples = recordLength * context.sampleRate;
+
+  await context.audioWorklet.addModule('./util/recorder/recorder-processor.js');
+
+  const recorder = new AudioWorkletNode(context, 'recorder', {
+    processorOptions: {
+      maxSamples,
+    },
+  });
+
+  let bufferResolve;
+  recorder.port.onmessage = (e) => {
+    if (e.data.message === 'RECORD_DONE') {
+      const channelBuffers = e.data.buffer;
+      const audioBuffer = new AudioBuffer({
+        length: maxSamples,
+        sampleRate: context.sampleRate,
+        numberOfChannels: channelBuffers.length,
+      });
+      channelBuffers.forEach((array, i) => audioBuffer.copyToChannel(array, i));
+
+      bufferResolve(audioBuffer);
+    }
+  };
+
+  const bufferPromise = new Promise((resolve) => {
```

Comment:
Here again, `bufferPromise` is difficult to understand, but I am out of ideas too. :)

Reply:
This promise will resolve with the recorded `audioBuffer` that is sent to the user.

Reply:
Yes. How about `RecordingCompletePromise`?

Reply:
So just to clarify the API, it would look like:
```js
const {recorder, RecordingCompletePromise} = await record(context, duration);
oscillator.connect(recorder).connect(context.destination);
// more code here etc...
const recordedChannelBuffers = await RecordingCompletePromise?;
//    ^Array<Float32Array>
```

Or would it be `recordingCompletePromise` instead?

Reply:
Yes. That looks great!

Yes. `recordingCompletePromise` SG.